core/primitives/tipset/tipset.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CPP_FILECOIN_CORE_PRIMITIVES_TIPSET_TIPSET_HPP
+#define CPP_FILECOIN_CORE_PRIMITIVES_TIPSET_TIPSET_HPP
+
+#include <boost/optional.hpp>
+#include "common/outcome.hpp"
+#include "common/outcome_throw.hpp"
+#include "primitives/block/block.hpp"
+#include "primitives/cid/cid.hpp"
+#include "primitives/ticket/ticket.hpp"
+#include "primitives/tipset/tipset_key.hpp"
+
+namespace fc::primitives::tipset {
+
+  enum class TipsetError : int {
+    NO_BLOCKS = 1,        // need to have at least one block to create tipset
+    MISMATCHING_HEIGHTS,  // cannot create tipset, mismatching blocks heights
+    MISMATCHING_PARENTS,  // cannot create tipset, mismatching block parents
+    TICKET_HAS_NO_VALUE,  // optional ticket is not initialized
+  };
+}
+
+/**
+ * @brief Outcome errors declaration
+ */
+OUTCOME_HPP_DECLARE_ERROR(fc::primitives::tipset, TipsetError);
+
+namespace fc::primitives::tipset {
+  /**
+   * @struct Tipset implemented according to
+   * https://github.com/filecoin-project/lotus/blob/6e94377469e49fa4e643f9204b6f46ef3cb3bf04/chain/types/tipset.go#L18
+   */
+  struct Tipset {
```

Comment:
It's weird that although it's a structure it has a lot of methods. Of course, they are getters so it's acceptable, but still low key unexpected.